README.md
# Toolbox.js

toolbox.js is a library to easily manage media components and perform simple data manipulations.

This library works as a standalone library, and can works well with any other JavaScript library like jQuery, jQueryMobile, Prototype, MooTools...

## Installation

You can download the library form GitHub repository.

| **File** | **Description** | **Size** | **Download** |
|:------------|:----------------|:---------|:-----------|
| `toolbox.js` | *uncompressed, with comments* | 6.87 KB | [Download](https://raw.github.com/Sevrahk/Toolbox/master/src/toolbox.js) |
| `toolbox.min.js` | *compressed, without comments* | 2.01 KB | [Download](https://raw.github.com/Sevrahk/Toolbox/master/src/toolbox.min.js) |

To use the library you just need to include it in your project and call it in your html documents.

``` html
<script src='toolbox.min.js'></script>
```

## Usage

To use the library you need to initialize a toolbox object.

### Example
```javascript
var toolbox = new Toolbox();
```

# Documentation

**Methods**

* Media Functions
    * [play](#play)
    * [pause](#pause)
    * [stop](#stop)
    * [resetMedia](#resetMedia)
    * [changeCurrentTime](#changeCurrentTime)
    * [isPlaying](#isPlaying)
    * [changeVolume](#changeVolume)
    * [isMedia](#isMedia)
* Array Functions
    * [shuffle](#shuffle)
    * [getMin](#getMin)
    * [getMax](#getMax)
    * [getSum](#getSum)
    * [getAvg](#getAvg)
* Utility Functions
    * [getRandom](#getRandom)
    * [degreesToRadians](#degreesToRadians)
    * [radiansToDegrees](#radiansToDegrees)

## Media Functions

<a name="play"></a>
### play(selector[, onEndCallback])

Play the selected media.  
`selector` string refers to the selected media tag or css class or ID.  
`onEndCallback` (optional) callback function.

```javascript
toolbox.play('.mediaTag');
```

```javascript
toolbox.play('.mediaTag', function() {
    console.log('Media ended');
});
```

<a name="pause"></a>
### pause(selector)

Set the media to pause.  
`selector` string refers to the selected media tag or css class or ID.

```javascript
toolbox.pause('.mediaTag');
```

<a name="stop"></a>
### stop(selector)

Stop the media and reset its current time.  
`selector` string refers to the selected media tag or css class or ID.

```javascript
toolbox.stop('.mediaTag');
```

<a name="resetMedia"></a>
### resetMedia(selector)

Change the media current time.  
`selector` string refers to the selected media tag or css class or ID.

```javascript
toolbox.reset('.mediaTag');
```

<a name="changeCurrentTime"></a>
### changeCurrentTime(selector, newTime)

Set the media current time to zero.  
`selector` string refers to the selected media tag or css class or ID.  
`newTime` nummber of seconds of the new current time.

```javascript
toolbox.changeCurrentTime('.mediaTag', 60);
```

<a name="isPlaying"></a>
### isPlaying(selector)

Check if the media is playing.  
`selector` string refers to the selected media tag or css class or ID.

```javascript
toolbox.isPlaying('.mediaTag');
```

<a name="changeVolume"></a>
### changeVolume(selector, newVolume[, duration, steps, onEndCallback])

Check if the media is playing.  
`selector` string refers to the selected media tag or css class or ID.  
`newVolume` number between 0 and 1.  
`duration` (optional) changing duration time in milliseconds.  
`steps` (optional) number of changing steps.  
`onEndCallback` (optional) callback function.

```javascript
toolbox.changeVolume('.mediaTag', 0.2);
toolbox.changeVolume('.mediaTag', 0.2, 10000, 10);
toolbox.changeVolume('.mediaTag', 0.2, 10000, 10, function() {console.log('Volume changed')});
```

<a name="isMedia"></a>
### isMedia(selector)

Check if the selector refers to an audio or a video tag.  
`selector` string refers to the selected media tag or css class or ID.

```javascript
toolbox.isMediaTag('.mediaTag');
```

## Array Functions

<a name="shuffle"></a>
### shuffle(array)

Shuffle the entered array.  
`array` the array to shuffle.

```javascript
var arr = [1, 2, 3, 4];
toolbox.shuffle(arr);
```

```javascript
toolbox.shuffle([1, 2, 3, 4]);
```

<a name="getMin"></a>
### getMin(array)

Get min number of the array.  
`array`

```javascript
var arr = [1, 2, 3, 4];
toolbox.getMin(arr);
```

```javascript
toolbox.getMin([1, 2, 3, 4]);
```

<a name="getMax"></a>
### getMax(array)

Get max number of the array.  
`array`

```javascript
var arr = [1, 2, 3, 4];
toolbox.getMax(arr);
```

```javascript
toolbox.getMax([1, 2, 3, 4]);
```

<a name="getSum"></a>
### getSum(array)

Get sum of numbers.  
`array`

```javascript
var arr = [1, 2, 3, 4];
toolbox.getSum(arr);
```

```javascript
toolbox.getSum([1, 2, 3, 4]);
```

<a name="getAvg"></a>
### getAvg(array)

Get avarage number.  
`array`

```javascript
var arr = [1, 2, 3, 4];
toolbox.getAvg(arr);
```

```javascript
toolbox.getAvg([1, 2, 3, 4]);
```

## Utility Functions

<a name="getRandom"></a>
### getRandom(min, max)

Get random value between entered values.  
`min` minimum (included).  
`max` maximum (included).

```javascript
toolbox.getRandom(1, 100);
```

<a name="degreesToRadians"></a>
### degreesToRadians(degrees)

Convert degrees to radians.  
`degrees` number to convert in radians.

```javascript
toolbox.degreesToRadians(100);
```

<a name="radiansToDegrees"></a>
### radiansToDegrees(radians)

Convert radians to degrees.  
`radians` number to convert in degrees.

```javascript
toolbox.radiansToDegrees(100);
```


CHANGELOG.md
# 1.0.0 / 2015-08-21
* First release


gulpfile.js
var gulp = require('gulp'),
    rimraf = require('rimraf'),
    jshint = require('gulp-jshint'),
    uglify = require('gulp-uglify'),
    rename = require('gulp-rename'),
    insert = require('gulp-insert'),
    p = require('./package.json')
    ;

//Folders and annotation
var src = 'src/',
    dest = 'dist/',
    annotation = '/* ' + p.name + ' v' + p.version + ' | Copyright (c) ' + new Date().getFullYear() + ' ' + p.homepage + ' | ' + p.license + ' license */\n';

gulp.task('clean', function(cb) {
    rimraf(dest, cb);
});

gulp.task('scripts', ['clean'], function() {
    return gulp.src(src + '*.js')
        .pipe(jshint())
        .pipe(jshint.reporter())
        .pipe(uglify())
        .pipe(rename('toolbox.min.js'))
        .pipe(insert.prepend(annotation))
        .pipe(gulp.dest(dest));
        ;
});

gulp.task('default', ['scripts'], function() {

});


src/toolbox.js
var Toolbox = (function() {
    'use strict';
    function Toolbox() {}

    //MEDIA
    /**
     * Play the selected media
     *
     * @example play('selector');
     * @example play('selector', function() {console.log('Media ended');});
     *
     * @param {String} selector
     * @param {Function} [onEndCallback]
     */
    Toolbox.prototype.play = function(selector, onEndCallback) {
        this.isMedia(selector);

        var mediaTag = document.querySelector(selector);
        mediaTag.play();

        if(typeof onEndCallback === 'function')
        {
            var callback = function() {
                onEndCallback();
                mediaTag.removeEventListener('ended', callback);
            };
            mediaTag.addEventListener('ended', callback);
        }
    };

    /**
     * Set the media to pause
     *
     * @param {String} selector
     */
    Toolbox.prototype.pause = function(selector) {
        this.isMedia(selector);

        document.querySelector(selector).pause();
    };

    /**
     * Stop the media and reset its current time
     *
     * @param {String} selector
     */
    Toolbox.prototype.stop = function(selector) {
        this.isMedia(selector);
        this.pause(selector);
        this.resetMedia(selector);
    };

    /**
     * Set the media current time to zero
     *
     * @param {String} selector
     */
    Toolbox.prototype.resetMedia = function(selector) {
        this.isMedia(selector);

        document.querySelector(selector).currentTime = 0;
    };

    /**
     * Change the media current time
     *
     * @param {String} selector
     * @param {number} newTime (seconds)
     */
    Toolbox.prototype.changeCurrentTime = function(selector, newTime) {
        this.isMedia(selector);

        document.querySelector(selector).currentTime = newTime;
    };

    /**
     * Check if the media is playing
     *
     * @param {String} selector
     * @returns {Boolean}
     */
    Toolbox.prototype.isPlaying = function(selector) {
        this.isMedia(selector);

        return !document.querySelector(selector).paused;
    };

    /**
     * Change the media volume
     *
     * @example changeVolume('selector', 0.2);
     * @example changeVolume('selector', 0.2, 10000, 10);
     * @example changeVolume('selector', 0.2, 10000, 10, function() {console.log('Volume changed')});
     *
     * @param {String} selector
     * @param {number} newVolume (between 0 and 1)
     * @param {String} [duration] : changing duration time in milliseconds
     * @param {String} [steps] : number of changing steps
     * @param {Function} [onEndCallback]
     */
    Toolbox.prototype.changeVolume = function(selector, newVolume, duration, steps, onEndCallback) {
        this.isMedia(selector);

        var mediaTag = document.querySelector(selector);
        if(duration === undefined || duration === null || duration === 0)
            mediaTag.volume = newVolume;
        else
        {
            var currentVolume = mediaTag.volume,
                volueIncreasing = mediaTag.volume <= newVolume,
                intervalTime = duration / steps,
                changeVolumeInterval;
            var volumeIncrement = (currentVolume - newVolume) * -1 / steps;

            changeVolumeInterval = setInterval(function() {
                currentVolume += volumeIncrement;
                if((volueIncreasing && mediaTag.volume < 1 && currentVolume <= newVolume) || (!volueIncreasing && mediaTag.volume > 0 && currentVolume >= newVolume))
                    mediaTag.volume = currentVolume;
                else
                {
                    clearInterval(changeVolumeInterval);
                    if(typeof onEndCallback === 'function')
                        return onEndCallback();
                }
            }, intervalTime);
        }
    };

    /**
     * Check if the selector refers to an audio or a video tag
     *
     * @param {String} selector
     * @throws {TypeError} : Selector doesn't exists or is not a media tag
     * @returns {Boolean}
     */
    Toolbox.prototype.isMedia = function(selector) {
        if(document.querySelector(selector) === null)
            throw new TypeError('The selector does not exist.');

        if(document.querySelector(selector).tagName.toLowerCase() !== 'audio' && document.querySelector(selector).tagName.toLowerCase() !== 'video')
            throw new TypeError('The selector does not refer to an audio or a video tag.');

        return true;
    };

    //ARRAY
    /**
     * Shuffle the entered array
     *
     * @param {Array} array
     * @returns {Array}
     */
    Toolbox.prototype.shuffle = function(array) {
        var currentIndex = array.length, tmp, randomIndex;
        while(currentIndex !== 0)
        {
            randomIndex = Math.floor(Math.random() * currentIndex);
            currentIndex -= 1;

            tmp = array[currentIndex];
            array[currentIndex] = array[randomIndex];
            array[randomIndex] = tmp;
        }

        return array;
    };

    /**
     * Get min number of the array
     *
     * @param {array} array
     *
     * @returns {number}
     */
    Toolbox.prototype.getMin = function(array) {
        return Math.min.apply(Math, array);
    };

    /**
     * Get max number of the array
     *
     * @param {array} array
     *
     * @returns {number}
     */
    Toolbox.prototype.getMax = function(array) {
        return Math.max.apply(Math, array);
    };

    /**
     * Get sum of numbers
     *
     * @param {array} array
     *
     * @returns {number}
     */
    Toolbox.prototype.getSum = function(array) {
        return array.reduce(function(prevVal, currVal) {
            return prevVal + currVal;
        });
    };

    /**
     * Get avarage number
     *
     * @param {array} array
     *
     * @returns {number}
     */
    Toolbox.prototype.getAvg = function(array) {
        return this.getSum(array) / array.length;
    };

    //UTILITY
    /**
     * Get random value between entered values
     *
     * @param {number} min (included)
     * @param {number} max (included)
     *
     * @returns {number}
     */
    Toolbox.prototype.getRandom = function(min, max) {
        return Math.floor(Math.random() * (max - min + 1)) + min;
    };

    /**
     * Convert degrees to radians
     *
     * @param {number} degrees
     * @returns {number} radians
     */
    Toolbox.prototype.degreesToRadians = function(degrees) {
        return degrees * Math.PI / 180;
    };

    /**
     * Convert radians to degrees
     *
     * @param {number} radians
     * @returns {number} degrees
     */
    Toolbox.prototype.radiansToDegrees = function(radians) {
        return radians * 180 / Math.PI;
    };

    return Toolbox;
})();
